segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/BaseMetadataStore.java

```diff
@@ -527,6 +542,158 @@ protected void setVersion(long version) {
         this.version.set(version);
     }
 
+    /**
+     * Helper class to encapsulates management of locks over group of keys.
+     */
+    static class MultiKeyLockManager {
+        /**
+         * Map to track active locks.
+         */
+        private final HashMap<String, LockData> keyToLockMap = new HashMap<>();
+
+        /**
+         * Creates a LockWrapper for reading.
+         * @param keysToLock Keys to lock.
+         * @return LockWrapper instance.
+         */
+        MultiKeyLock readLock(String... keysToLock) {
+            return new MultiKeyLock(this, true, keysToLock);
+        }
+
+        /**
+         * Creates a LockWrapper for writing.
+         * @param keysToLock Keys to lock.
+         * @return LockWrapper instance.
+         */
+        MultiKeyLock writeLock(String... keysToLock) {
+            return new MultiKeyLock(this, false, keysToLock);
+        }
+
+        private LockData addReference(String key) {
+            LockData lockData;
+            synchronized (keyToLockMap) {
```

Comment:
If we use a ConcurrentHashMap this block can be expressed as a 'compute' operation,l with the main benefit that we will increase parallelism because in this case the lock is per key (actually per bucket).
With the current approach you have a single global lock

Reply:
One problem with that approach is that we are also keeping track of reference counts so that we can efficiently delete unused keys.  Now we could use AtomicInteger for atomic update to the count. But the problem is that removeReference is not atomic anymore. Between decrementing + checking counter value and actually deleting it , another thread could increment reference count and that causes key being deleted incorrectly.

Let me see if double checking solves the issue
Assuming count is AtomicInteger
```
if (0== count.decrementAndGet()) {
    synchronized {
          if (0== count.decrementAndGet()) {
              remove()...
          }
    }
}
```
But otherwise we'll have to build something more complicated. We'll do that only if absolutely necessary.

Reply:
You can use ConcurrentHashMap.merge, if the computation returns null the entry is remove by the map (IIRC)
see
https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#merge-K-V-java.util.function.BiFunction-

Reply:
no need for an AtomicInteger

Reply:
If the key is present and multiple threads access it then we still need to keep track of ref counter. That operation needs to be atomic or we'll have some leaks or worse - premature deletion. 

Enter -
1. Get Value for key, if not present insert.
2. Increase ref count.

Leave 
1. Get Value for key.
2. Decrease ref count, if 0 then delete key.

Thread 1 :  GetValue , doesn't find insert
Thread 2 :  Get Value, finds existing value
Thread 2:  Increment counter
Thread 2: Done . Decrement counter
Thread 2 : Delete key as count ==0;
**Thread 1 : Dealing with value that  is not present in map 
**Thread 3: .GetValue , doesn't find insert...essentially new threads will use new locks not existing one. The key is no more protected against concurrent access.
** Thread 1 : Null Exceptions when releasing 

I couldn't figure out an elegant way to achieve this without use of synchronized block in both addReference and releaseReference. 
So decided to do the simplest possible thing .

Reply:
This is what we have in HerdDB

https://github.com/diennea/herddb/blob/c44c80d8326d2280d207bfa5447f3db4d024742c/herddb-utils/src/main/java/herddb/utils/LocalLockManager.java#L65

Reply:
Got it. Thanks
I'll try that.

Reply:
I tried that approach but didn't quite get it to work as expected. I got into some weird race conditions that I couldn't reason with (eg Unexpected releasing of non-existent key).  

While synchronize is slower, I think for this specific purpose it is "good enough".
This code is going to be rewritten to be fully async in 0.9 ...so I do not want to invest too much in this part.

Reply:
Ok.
Better to have something that work now.

Thanks for checking